### PR TITLE
Bugfix/2400 change error message repo with new storage type

### PR DESCRIPTION
**Pull request by lynnro314**
Close: #2400 

When trying to create a repository with a different storage type than the one defined in the configuration, the error message was not descriptive.
I made two changes:

1. In create repo I added a verification of whether the storage type is the same as the one defined in the configuration. In this case, a descriptive error message will be printed.
2. I changed the error message in the UI to be descriptive.

### CODE REVIEW

File: pkg/api/controller.go
Comment (arielshaqed) [inline pkg/api/controller.go:1107]: Not sure why this is not part of the ensure* call. I assume we'd want the same error detection if we added another "create repo" operation. 
If the issue is better error reporting from that call then we might want to do it there. E.g. Return an internal error code and an external message.

File: webui/src/lib/components/repositoryCreateForm.jsx
Comment (arielshaqed) [inline webui/src/lib/components/repositoryCreateForm.jsx:81]: Can we use an error message from the server here?

**Review by arielshaqed — APPROVED** [review]
Neat, and *very* helpful to our users! 

I  think we might improve the structure of where we do the test in the API server and how much the client needs to know in order to report the error.

Also can we treat this somehow in `api_test.go`?

Thanks!

File: pkg/api/controller.go
Comment (talSofer) [reply pkg/api/controller.go:1107]: Agree with Ariel, this logic belongs to the ensureStorageNamespaceRW method because it's part of ensuring the validity of the storage namespace. Can you please move it there?

File: pkg/api/controller.go
Comment (talSofer) [inline pkg/api/controller.go:1109]:
```suggestion
		writeError(w, http.StatusBadRequest, "error parsing storage namespace. Please enter a valid repository url.")
```

File: pkg/api/controller.go
Comment (talSofer) [reply pkg/api/controller.go:1107]: nit; it's up to you but I prefer the name parsedNamespace for readability reasons

File: pkg/api/controller.go
Comment (talSofer) [inline pkg/api/controller.go:1114]:
```suggestion
		writeError(w, http.StatusBadRequest, "error creating repository: invalid storage type.")
```

File: pkg/api/controller.go
Comment (talSofer) [inline pkg/api/controller.go:1117]: It is not clear what "rel" stands for.  can you please rename this variable? 
A suggestion: 
Use `adapterStorageType` instead of `relStorageType`, and `namespaceStorageType` instead of `storageType`

Comment (talSofer) [reply pkg/api/controller.go:1117]: Also, it is probably most readable to get rid of the variable and do: 
`if c.BlockAdapter.BlockstoreType() != namespaceStorageType.BlockstoreType()`

**Review by talSofer — CHANGES_REQUESTED** [review]
A great improvement to our error reporting, thanks!

File: pkg/api/controller.go
Comment (lynnro314) [reply pkg/api/controller.go:1117]: No problem, will rename it. Also, I added the variable since I'm using it later when I'm writing the error.

File: pkg/api/controller.go
Comment (arielshaqed) [inline pkg/api/controller.go:1117]: Yikes... this can leak potentially sensitive information.  A good way to handle something like this is to return _two_ values (and document them!) from `ensureStorageNamespace`: the first is an `error`, the second is a `string` which will be shown to the user.  Then it is easier to ensure that nobody ever returns sensitive information from that function.

File: pkg/block/namespace.go
Comment (arielshaqed) [inline pkg/block/namespace.go:129]: Generally the negative is already in the base error message, so it may make sense merely to describe the failure:
```suggestion
		return QualifiedKey{}, fmt.Errorf("parse URI: %w", err)
```

File: webui/src/lib/components/repositoryCreateForm.jsx
Comment (arielshaqed) [reply webui/src/lib/components/repositoryCreateForm.jsx:81]: I believe that this is still open.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Must-have change is something to make it less likely to leak sensitive information on error messages written out to clients.

**Review by arielshaqed — APPROVED** [review]
After talking f2f with @lynnro314 I agree that my comments about a _potential_ leak from an error message does not exist today and is anyway possible through _every other_ block adapter method.  So no justification to block on this.
Approving and we shall open a **separate** issue to return separate messages to users from _all_ API and S3 gateway calls.

File: pkg/api/controller_test.go
Comment (talSofer) [inline pkg/api/controller_test.go:328]: This part looks redundant. You only want to create a single repo with a conflicting storage namespace (as you did in line 329)

File: pkg/api/controller_test.go
Comment (talSofer) [inline pkg/api/controller_test.go:324]: nit; you can use the word "conflicting" instead of "wrong"?  wrong is less descriptive

**Review by talSofer — APPROVED** [review]
Looks good! please address the comments before you merge the code

File: pkg/api/controller_test.go
Comment (lynnro314) [reply pkg/api/controller_test.go:328]: I want first to create a repo with some storage type, and then create another repo with different storage type. Sounds Ok?

Comment (talSofer) [reply pkg/api/controller_test.go:328]: Why do you need to create the first repo? Do you mean that you are trying to test both the valid and invalid cases, or you meant something else? 

The sub-test description says "create repo with conflicting storage type", so what I expect to see is an attempt to create a repository with an invalid storage namespace that fails. 

If you are trying to test both cases I would separate their test. that is, you can define another test case with the description "create repo with a storage type" but this is already tested in line 289.

File: pkg/api/errors.go
Comment (nopcoder) [inline pkg/api/errors.go:10]: I think that we didn't fail to ensure - we failed to find a matched storage namespace.
Error like - mismatch storage namespace or something like it describes the issue.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/api/controller_test.go
Comment (lynnro314) [reply pkg/api/controller_test.go:328]: You are right, this test is not necessary so I took it down.

File: pkg/api/auth_middleware.go
Comment (nopcoder) [inline pkg/api/auth_middleware.go:125]: not your code - but can you add WithError to log the information

File: pkg/api/controller.go
Comment (nopcoder) [inline pkg/api/controller.go:1121]:
```go
		var ErrFailedToAccessStorage = " failed to access storage"
```

```suggestion
		var retErr error
		var e *url.Error
		if errors.As(err, &e) && e.Op == "parse" {
			retErr = err
		} else if errors.Is(err, block.ErrInvalidNamespace) {
			retErr = fmt.Errorf("repository storage type must match: %s", c.BlockAdapter.BlockstoreType())
		} else {
			retErr = ErrFailedToAccessStorage
		}
```

File: pkg/api/controller.go
Comment (nopcoder) [inline pkg/api/controller.go:1126]:
```suggestion
		writeError(w, http.StatusBadRequest, fmt.Errorf("failed to create repository: %w", retErr))
```

**Review by nopcoder — APPROVED** [review]
Looks good

**Pull request merged**
